### DISCUSSION TITLE
Interaksjonsutfordringer i datepicker

### DISCUSSION THREAD

**Discussion by joms** [fremtind/jokul#2436] [Annet]
Det kan se ut som det er flere interaksjonsutfordringer i datepickeren i dag. Den ene problemstillingen er ren interaksjon, og den andre går på tastaturnavigasjon:

## Interaksjon

Når man navigerer i datepickeren ved å bare sette årstall eller år, for så å klikke utenfor datepickeren oppdateres ikke state. Her tror jeg det ligger en forventning hos de fleste om at man oppdaterer datoen.

Ett forslag for å løse denne problemstillingen er å fjerne "selected" staten på datoen, sånn at man må velge dato selv igjen. Dette gjør at man kan "avbryte" datoendringen også, om man skulle ønske dette.

## Tastaturnavigasjon

Når man navigerer i datepickeren ved å bare sette årstall eller år med tastaturet, burde kan kunne trykke enter fra hvor som helst for å oppdatere state så lenge datoen har en selected state.

Men om man velger en løsning hvor selected state blir fjernet, og man i stedet bare sitter igjen med fokus må det sikres at man beholder en fokus state som kan brukes til tastaturnavigasjon. F.eks. ved å fjerne bakgrunnsfargen på dato og heller bruke en fokusring for å vise hvor tastaturnavigasjon starter.

**Comment by joms** [comment]
Eksempel fra Lou på hvordan det kan løses på en god måte: https://www.sbb.ch/de/fahrplan.html

**Comment by onordbo** [comment]
Jeg ser ikke for meg noen naturlig grunn til at vi skal divergere fra det som i mine øyne hvertfall er "standard praksis" for datepicker. Brukerne burde ta et aktivt valg om å velge en ny dato inne i kalendermåneden, hvis de har navigert ut av måneden den valgte datoen står i, uavhengig av om de bruker mus eller tastatur. Støtter forslaget ditt om å la "valgt dato" forbli der den var definert før man naivgerer vekk @joms. +1 for tastaturnavigasjon med en bestrøket fokusring uten bakgrunn.

**Comment by piofinn** [comment]
Om vi skal gjøre denne endringen (interaksjonsendringen) bør vi nok gå bort fra Core Datepicker og lage vår egen kalender fra scratch. Det er nok ikke en dum idé uansett, så mange endringer og workarounds vi har gjort med den som det er allerede 😅

"Navigasjonen" i Core Datepicker går ut på å faktisk endre datoen (det er derfor den markeres hos oss), og så vidt jeg kan se er det ingen andre muligheter lagt inn der. Det er litt spesielt, men den er laget slik at kalenderen bare er en av mange måter du kan velge dato på. Selv om vi skulle klare å finne en workaround også for dette begynner det å likne på en [Thesevs' skip](https://en.wikipedia.org/wiki/Ship_of_Theseus)-situasjon, og da er det kanskje greit å eie hele kodegrunnlaget selv?

**Reply by wkillerud** [reply]
Ja, den patchen vi gjør er kjip å få med hver gang det oppgraderes. Det er nok mye å lære fra å se på både Core Datepicker og andre, men +1 på å eie hele kodebasen selv.

**Reply by wkillerud** [reply]
Opplevde også at dagens DatePicker var litt kronglete å bruke med `react-hook-form` i #2489. Tror vi må gruble litt over hvordan DX skal være også – hvor mye som gjøres automagisk inni komponenten, hva som skal være input og output.

**Reply by joms** [reply]
https://github.com/its-danny/use-lilius Bare kaster inn denne her så vi husker den. Jeg har lagd en datepicker i et annet prosjekt som bruker denne, og det funker veldig bra.
